Add brief documentation of Cython binding and contribution guidelines (#268)
VeriFIT/mata: bindings/python/CONTRIBUTING.md

  * [Calling functions of shared pointer object](#calling-functions-of-shared-pointer-object)
  * [Supporting Python for loops](#supporting-python-for-loops)
  * [Working with strings](#working-with-strings)
  * [Helping Cython to infer types](#helping-cython-to-infer-types)
  * [Using pre- and post-increments (`var++` and `++var`)](#using-pre--and-post-increments-var-and-var)
  * [Working with exceptions](#working-with-exceptions)
* [Importing between modules](#importing-between-modules)
  * [Importing C/C++ defined functions and symbols](#importing-cc-defined-functions-and-symbols)
  * [Importing Python defined functions and symbols](#importing-python-defined-functions-and-symbols)
  * [Importing standard containers and types](#importing-standard-containers-and-types)
* [Troubleshooting](#troubleshooting)
  * [An error when importing Python object from other module](#an-error-when-importing-python-object-from-other-module)
  * [A weird error when Cython cannot retype C/C++ object to Python object](#a-weird-error-when-cython-cannot-retype-cc-object-to-python-object)
* [Quick glossary](#quick-glossary)
<!-- TOC -->

## File/Directory Structure

There are two types of files:

> - `*.pxd`: Cython files containing declarations/definitions of the C/C++ API.

Reviewer: Append: "specifying what you want to import from the C/C++ API"

Reviewer: PS: More comments to come. Experimenting with external PR review workflow.

Author: Sure. Btw. for short rewording or grammar fixes, please, prefer suggestions, so I can directly commit the changes whenever possible.

    - `alphabets.*`: Wrappers for `Mata::Alphabet` members and related structures or algorithms.
    - `parser.*`: Wrappers for `Mata::Parser` members and related structures or algorithms.
    - `plotting.*`: Helper functions for displaying `libmata` members in Python code, interpreters, Jupyter notebooks, etc.
    - `utils.*`: Wrappers for `Mata::Utils` members and related structures or algorithms.
- `tests/`: Tests for the Cython wrapper.
    - `conftest.py`: Global configuration of tests, containing fixtures, setups, etc.
- `Manifest.in`: Specification of the contents of the wrapper distribution for PyPI.
- `pyproject.toml`: Specification for the wrapper distribution for PyPI.
- `requirements.txt`: Specification of the wrapper's requirements and dependencies. Run `make init` or `pip3 install -r requirements.txt` to install them.
- `setup.py`: Specification of the build process.

## Adding a new function

  1. Declare the function in `.pxd` file:
    
     - For a class function, declare it within the class definition:

```cython
cdef extern from "mata/nfa.hh" namespace "Mata::Nfa":
#                ^-- source file          ^-- namespace
    cdef cppclass CNfa "Mata::Nfa::Nfa":

>     #             ^-- class name and its full C/C++ name

Reviewer: Split pointer to class name and full C/C++ name

It may require writing the explanation in two lines.
```suggestion
    cdef cppclass CNfa "Mata::Nfa::Nfa":
    #             ^-- class name in binding
    #                  ^-- its corresponding full C/C++ name
```

        void make_initial(State)
        #    ^- signature of class function
```

>      - For a non-class function, declare it under its corresponding environment:

Reviewer: This will not display properly as a bullet point. Instead, this displays as another code block.

```cython
cdef extern from "mata/nfa-plumbing.hh" namespace "Mata::Nfa::Plumbing":
#                ^-- source file                  ^-- namespace
    cdef void get_elements(StateSet*, CBoolVector)

>     #    ^-- signature

Reviewer: ```suggestion
    #    ^-- function signature using previously defined 
    #         Cython types from C/C++ API as parameters
```

```

>      - You need to specify: (1) the file where the function is declared (e.g., nfa-plumbing.hh), 

Reviewer: This also displays wrongly as a code block.

    #        ^-- call to wrapped functions
    return { state for state in c_states }
    #      ^-- conversion to Python set
```

## Adding a new module

  1. Create a `module.pxd` file and declare the functions from libmata that you want to use in the binding.
  2. Create a `module.pyx` file, where you will `cimport module` and write the implementation for functions calling the declared functions as `module.declared_function(...)`.
  3. Add the module to the list of extensions in `setup.py`:

```cython
extensions = [
    Extension(
        f"libmata.{pkg}",
        sources=[f"libmata{os.sep}{pkg.replace('.', os.sep)}.pyx"] + project_sources,
        include_dirs=project_includes,
        language="c++",
        extra_compile_args=["-std=c++20", "-DNO_THROW_DISPATCHER"],
    ) for pkg in (
        'nfa.nfa', 'alphabets', 'utils', 'parser', 'nfa.strings', 'plotting', 'module'
    )
]
```

## Adding a new class

  1. Declare the class in the `.pxd` file

```cython
cdef extern from "mata/alphabet.hh" namespace "Mata":
#                ^-- source file              ^-- namespace
    cdef cppclass COnTheFlyAlphabet "Mata::OnTheFlyAlphabet" (CAlphabet):
        

>         #         ^-- name in bind  ^-- full name           ^-- inheritance

Reviewer: Clarify the description.
```suggestion
    cdef cppclass COnTheFlyAlphabet "Mata::OnTheFlyAlphabet" (CAlphabet):
        
        #         ^-- name in binding
        #                           ^-- full name in C/C++ API
        #                                                    ^-- inheritance using
        #  previously defined Cython classes 
```

        StringToSymbolMap symbol_map

>         # ^-- type        ^--- property name

Reviewer: ```suggestion
        StringToSymbolMap symbol_map
        # ^-- previously defined Cython type
        #                 ^--- property name
```

        COnTheFlyAlphabet(StringToSymbolMap) except +
        # ^-- constructor                    ^-- can fire exception
        StringToSymbolMap get_symbol_map()
        # ^-- class method
```

  2. In `pyx` file declare the Python object that will serve as a wrapper:

```cython
cdef class OnTheFlyAlphabet(Alphabet):
    pass
```

  3. In the wrapper object define the variable that holds pointer to the original C/C++ object.

```cython
cdef class OnTheFlyAlphabet(Alphabet):
    cdef COnTheFlyAlphabet *thisptr
    #    ^-- pointer to OnTheFlyAlphabet in the C/C++ library
```

  4. Create `__cinit__` and `__dealloc__` functions which are created when constructing new Python object.

```cython
cdef class OnTheFlyAlphabet(Alphabet):
    cdef COnTheFlyAlphabet *thisptr

    def __cinit__(self, State initial_symbol = 0):
    # ^-- constructor function
        self.thisptr = new COnTheFlyAlphabet(initial_symbol)
    #   ^-- populates the pointer through visible constructors and `new` keyword

    def __dealloc__(self):
    # ^-- destructor function
        del self.thisptr
```

>   5. You can now define your class as `OnTheFlyAlphabet()` in Python code

Reviewer: Add missing full stop.
```suggestion
  5. You can now define your class as `OnTheFlyAlphabet()` in Python code.
```

```

* Second, implement the corresponding Python operators in `pyx` file.

```cython
# @file: nfa.pyx
cdef class Move:
    def __lt__(self, Move other):
        return dereference(self.thisptr) < dereference(other.thisptr)

    def __gt__(self, Move other):
        return dereference(self.thisptr) > dereference(other.thisptr)

    def __le__(self, Move other):
        return dereference(self.thisptr) <= dereference(other.thisptr)

    def __ge__(self, Move other):
        return dereference(self.thisptr) >= dereference(other.thisptr)
```

* Now, you can compare objects in Python.

# Other advanced features

## Dereferencing and retyping the object

* **Problem**: You have a pointer (e.g., returned from function) and need to dereference it and 
  retype as well, as Cython cannot infer the type.
* **Solution**: You `cimport` the `dereference` operator and then explicitly retype using `<CType>` notation.

```cython
from cython.operator import dereference
# ^-- import from cython.operator the implementation of dereference
<CAlphabet>dereference(alphabet.as_base())
#                       ^-- the function returns CAlphabet* type
#           ^-- dereference returns CAlphabet

> # ^-- the topmost function helps Cython to know, it is CAlphabet

Reviewer: ```suggestion
# ^-- the topmost function helps Cython to infer that it is CAlphabet
```
Rephrase the description.

```

* Note: You might need to use parenthesis to restrict the scope of the the retype, i.e. `(<CType>...)`.

### Dereferencing shared pointer

* **Problem**: You want to dereference shared pointer in the `pyx` code.
* **Solution**: You simply call `.get()` function and then you can safely dereference the pointer.

```cython
from cython.operator import dereference
dereference(rhs.thisptr.get())
# ^-- cython implementation of dereference (operator*)
#           ^-- rhs object holds variable `shared_ptr[T] thisptr`
#                       ^-- get() returns T*
```

### Dereferencing iterator

* **Problem**: You want to access the value of the iterator in `pyx` code.
* **Solution**: You simply dereference the iterator.

```cython
from cython.operator import dereference
dereference(it).symbol
#           ^-- it holds some iterator object
```

## Using null pointer

* **Problem**: You want to use null pointer in your `pyx` code.
* **Solution**: You simply use `NULL` constant, that does not need to be imported and is supported

>   by Cython. There seems to be no support for `nullptr`, though.

Reviewer: Remove unnecessary explanation.
```suggestion
  by Cython. `nullptr` is trivially replaced with C-style `NULL`
```

    trans.append(Trans(c_transition.src, c_transition.symb, c_transition.tgt))
```

## Working with strings

* **Problem**: You want to work with strings and pass them between binding and wrapped library.
* **Solution**: You need to use `encode()` when passing strings to C/C++ functions, and `decode()`,
  when recieving results.

```cython
k.encode('utf-8')
# ^- when passing strings to C/C++ library you need to encode
result.decode('utf-8')
# ^- when retrieving results from C/C++ you need to decode
```

## Helping Cython to infer types

* **Problem**: Cython fails to infer types for the variables and cannot compile the binding.
* **Solution**: You need to help Cython a little: add types to parameters, add types to cdefined 
  variables, or manually retype the variable.

```cython
(<mata_nfa.Nfa>noodle_segment).thisptr = c_noodle_segment
# ^-- tell Cython that this variable is of given type
objects: List[subprocess.Popen] = []
# ^-- tell Cython the type of the Python object
cdef mata_nfa.stringstream* output_stream
# ^-- define variable as C/C++ type
def union(cls, Nfa lhs, Nfa rhs):
    pass
#              ^-- tell Cython types of parameters
```

> ## Using pre- and post-increments (`var++` and `++var`)

Reviewer: Move this section above section `Working with iterators` where `postincrement` is already used.

* **Problem**: You want to manually use pre/post increments in your code.
* **Solution**: You have to import  `postincrement` and `preincrement` from `cython.operator`.

```cython
from cython.operator import postincrement as postinc, preincrement as preinc
postinc(it)
preinc(iterator)
```

## Working with exceptions

* **Problem**: Your function can fire exceptions.
* **Solution**: You have to tell Cython how to handle these.

```cython
void add(CTrans) except -1
#                ^-- exceptions are returned as -1 in Cython
void add(CTrans) except +
#                ^-- function may fire exception; C/C++ exceptions will be translated to Python exceptions
void add(CTrans) except +MemoryError
#                ^-- function may fire MemoryError (corresponds to bad_alloc) 
void add(CTrans) noexcept +
#                ^-- function will not fire exception
```

* Note, there are other ways of handing exceptions, but these should suffice.

> # Importing between modules

Reviewer: Maybe start the whole guide with this section?

Reviewer: On second thought, you explain a bit of importing at the start. This should be sufficient. I will close this conversation.
